controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java

```diff
@@ -907,15 +939,35 @@ private HttpResponse deploy(String tenantName, String applicationName, String in
                                                                  vespaVersion,
                                                                  deployOptions.field("ignoreValidationErrors").asBool(),
                                                                  deployOptions.field("deployCurrentVersion").asBool());
+
+
         ActivateResult result = controller.applications().deploy(applicationId,
                                                                  zone,
                                                                  applicationPackage,
                                                                  applicationVersion,
                                                                  deployOptionsJsonClass,
                                                                  Optional.of(getUserPrincipal(request).getIdentity()));
+
         return new SlimeJsonResponse(toSlime(result));
     }
 
+    /** Find the minimum value of a version field in a zone */
+    private Version wantedSystemVersion(ZoneId zone, SystemApplication application) {
```

Comment:
Would be nice to have this in the `Controller` instead, if it's the same as the code already used by the `InfrastructureUpgrader`. 

(Also, clean up imports at the top :))

Reply:
Or just leave it at the system version. Is there any use case for triggering an early upgrade?

Reply:
I thought the idea was to avoid premature upgrade. Will the routing application be upgraded before system version is set?

Reply:
Yes, when the controller upgrades the system applications (hosts, config servers and zone app) are gradually upgraded to the controller version. Once all system applications have upgraded, the system version is bumped.

I think it's best to keep it on the system version then (e.g. just use `Controller#systemVersion`).

Reply:
So using the system version could cause a downgrade then?

Reply:
If you happen to collide with a system upgrade, then yes, it could happen. I think the safest way of guaranteeing no collisions is to reject deployments of zone app when system version != controller version. This avoids second-guessing the orchestration.

Reply:
fixed